Jeff Mahoney <jeffm@suse.de>:

fsstress manages to setup a sequence of events that allow an attempt to
perform direct-io on a tail. bh_result->b_page == NULL, which causes the
PageLocked and PageWriteback checks to perform a NULL deref, causing the
oops.

diff -ruN -X dontdiff linux-2.6.3.2/fs/reiserfs/inode.c linux-2.6.3/fs/reiserfs/inode.c
--- linux-2.6.3/fs/reiserfs/inode.c	2004-02-23 16:31:57.128850200 -0500
+++ linux-2.6.3.2/fs/reiserfs/inode.c	2004-02-19 11:24:44.000000000 -0500
@@ -453,7 +453,7 @@
         /* make sure future calls to the direct io funcs for this offset
         ** in the file fail by unmapping the buffer
         */
-        reiserfs_unmap_buffer(bh_result);
+        clear_buffer_mapped(bh_result);
         ret = -EINVAL ;
     }
     /* Possible unpacked tail. Flush the data before pages have
iff -ruN -X dontdiff linux-2.6.3.2/fs/reiserfs/journal.c linux-2.6.3/fs/reiserfs/journal.c
